### ISSUE TITLE
[Site Request] Beta Crunchyroll

### ISSUE THREAD

**Issue by federicorosso1993** [yt-dlp/yt-dlp#915]
## Checklist

- [x] I'm reporting a new site support request
- [x] I've verified that I'm running yt-dlp version **2021.09.02**
- [x] I've checked that all provided URLs are alive and playable in a browser
- [x] I've checked that none of provided URLs violate any copyrights
- [x] The provided URLs do not contain any DRM to the best of my knowledge
- [x] I've searched the bugtracker for similar site support requests including closed ones

## Example URLs

https://beta.crunchyroll.com/it/watch/G8WUN2ND8/La-citt-delle-bestie-magiche-coperta-di-neve
https://beta.crunchyroll.com/watch/GJWU2J7Z8/Clash-Class-A-vs-Class-B

## Description

I opened this issues on the youtube-dl main repo on 13 Jun
https://github.com/ytdl-org/youtube-dl/issues/29290

right now crunchyroll only work as a beta in the US (you cannot exit the beta) and you can also use the beta in the rest of the world (but you can still exit it).
Also @hyugogirubato made a little python script that implement most of the thing yt-dlp should implement to make beta crunchyroll work (he also implemented 100% the new API from log in to download thanks to a scrapping of the mobile app)
https://github.com/hyugogirubato/Kamyroll-Python

and I made my own implementation of that in C# (my version only use basic API call to make it possible to download from an episode url, and I mapped the API to C# POCO class. Also I made this based on the old version of Kamyroll-python and it use the log in based on the device id of the crunchyroll desktop app and I made a comment for the log in based on the web log in)
https://github.com/federicorosso1993/Kamyroll-CSharp

Both version are working with the new beta crunchyroll and should help to implement beta crunchyroll support to yt-dlp.
Not just that, youtube-dl and yt-dlp crunchyroll implementation (not beta) are already partially broken, for example you cannot use username and password to log in to the website, you need to use a cookie file with the user agent where it was extracted... and you need to do that on a real browser. That's not needed like you can see in both my and the python implementation (even if you don't use the scrapping of the mobile app (for any reason) you can use the a code like this var cookie = cookieContainer.GetCookies(uri).Cast<Cookie>().FirstOrDefault(x => x.Name == "etp_rt"); in C# and a code like this r = requests.get(endpoint)
    etp_rt = r.cookies.get("etp_rt") in python to get the needed cookies (for reference to this solution look at my c# version or the old version of Kamyroll-python https://github.com/hyugogirubato/Kamyroll-Python/blob/ffba19ec6f8c81dae4e5f35e368748fe2fca6494/kamyroll.py, of course I suggest you to just implement it like in the last version of Kamyroll-Python where you don't even need cookies but I think the old implementation should be kept to get an alternative to the mobile API that's also why I did not update my own implementation, I would like to avoid things that I don't understand like the scrapping of the mobile app, I'll still update it as a comment once I got some time).
There is no real reason to fix the old crunchyroll implementation if the beta works already great so I think it's time to add it to yt-dlp.

Like I said in the old issues I'm not a python expert so I don't really know how to change the yt-dlp code (I would have done a pull request to youtube-dl months ago since I made the C# version if I was able to understand how to do it in python), I was able to read and understand the old version of Kamyroll-Python thanks to @hyugogirubato help and I will keep update my own version if I can understand what to change, but I would like a big repo like yt-dlp to implement it too so that if I or hyugogirubato will not be able to understand how to update it there will be many capable developers to fix it.

**Comment by Ashish0804** [comment]
Weird
I can't use the site even with a vpn (IT)
It redirects me to `https://www.crunchyroll.com/`
I wonder if it is bcoz of account

If @federicorosso1993 can u access the links u provided without any account?

**Comment by federicorosso1993** [comment]
> Weird
> I can't use the site even with a vpn (IT)
> It redirects me to `https://www.crunchyroll.com/`
> I wonder if it is bcoz of account
> 
> If @federicorosso1993 can u access the links u provided without any account?

Sorry I forgot that you can only access beta crunchyroll with an account (even a free one should be ok in the US).
The api that we used can also be used without an account that's why I forgot.

**Comment by Ashish0804** [comment]
you mean you can get the videos via API without any account?

**Comment by federicorosso1993** [comment]
> you mean you can get the videos via API without any account?

Yes.

**Comment by Ashish0804** [comment]
why do you need login feature then?
Does using account has any benefits here?

**Comment by federicorosso1993** [comment]
> why do you need login feature then?
> Does using account has any benefits here?

Yes. If you got a premium account you can access simulcast one week before release for free user. And there are some videos that are premium only

**Comment by Ashish0804** [comment]
And these features will work even with a free account?

**Comment by federicorosso1993** [comment]
> And these features will work even with a free account?

No, that's why this will need a log in feature.

**Comment by Ashish0804** [comment]
you will need to provide a paid account then to the dev to implement this feature (Assuming he doesn't have it already)
Else all the rest of features can be implemented without writing any login function

**Comment by federicorosso1993** [comment]
> you will need to provide a paid account then to the dev to implement this feature (Assuming he doesn't have it already)
> Else all the rest of features can be implemented without writing any login function

You don't need a paid account to implement the API. Just use a free account and you can access the beta website and also you can try the login API with that account. Also the code is already there in the Kamyroll-python version, I don't know how to integrate it with yt-dlp but you don't even need to create anything different

**Comment by Ashish0804** [comment]
I can't create a free account (if by free u mean the 14 days trial one) without providing my CC info.

Account is needed to test those features, I doubt anyone will add those without testing.

**Comment by federicorosso1993** [comment]
> I can't create a free account (if by free u mean the 14 days trial one) without providing my CC info.
> 
> Account is needed to test those features, I doubt anyone will add those without testing.

No there is a free account

https://www.crunchyroll.com/login?next=%2Fit
